### DISCUSSION TITLE
Sensor class for confidence score (% or fraction)

### DISCUSSION THREAD

**Discussion by kiler129** [home-assistant/architecture#874] [Entity Models] [answered]
### Use-case
With the current raise in popularity of various ML models some of their results will land in HA. A common example would be object detection performed by a camera. An real object detection example from a smart camera may contain information like below:
```json
{
  "detections": [
    {
      //...
      "className": "car",
      "score": 0.83984375,
      "id": "2.0",
      "history": {
        "firstSeen": 1676069196843,
        "lastSeen": 1676069272128
      },
      "zones": []
    }
  ],
  //...
}
```

### Problem with the current model
The results of e.g. object detection vary greatly. Thus, exposing just an information _"package detected"_ as a binary sensor may lead to flaky automations.

One of the solutions, possible today, is to allow users to specify thresholds in an integration for e.g. a camera, and let the integration code decide whether a given detection passes the threshold. However, this is flawed in many ways. Detection sensitivity of the currently available models vary greatly between day and night and different thresholds are needed in the real world depending on the luminance levels or IR blaster activity. In addition, letting integrations directly compare thresholds doesn't allow for more advanced scenarios like `if >0.5 for 5 minutes or >0.9 any duration`.

### A better solution
 The solution is to expose the confidence score as a sensor value. This lets the user fine-tune their automations to their environment. Currently, the `Sensor` entity model does not contain any class which will be even remotely usable for the purpose. I'm suggesting to add a new class for that purpose. While usually confidence scores are usually presented as a number between 0 and 1, I believe that HA can standardize to percents to simplify user's understanding and thus improving UX.

### Scope
To my knowledge, currently there are no integrations exposing object detection from cameras in core. I'm currently working on one. However, the usability of confidence score goes beyond just object detection from cameras, i.e. it's a pretty universal entry.

### Suggested shape
- Constant: `SensorDeviceClass.CONFIDENCE_SCORE`
- Supported Units: `%`
- Description: `Chance of the another measurement's value being correct`

**Comment by parautenbach** [comment]
To give a real example where this is already used: the HA companion app exposes an activity sensor for iPhones. This sensor contains a confidence attribute (not as a percentage, but using the labels high, medium and low). This comes directly from iOS based on the Bionic chip and Core ML APIs.

**Reply by kiler129** [reply]
Adding to that, the CoreML actually provides the exact values - https://developer.apple.com/documentation/vision/vnobservation/2867220-confidence

**Comment by frenck** [comment, marked as answer]
I've discussed this proposal today in an architecture meeting, and we have concluded that there is not enough traction to support this right now. Therefore, I'll go ahead and close this one for now.

../Frenck
